onnxruntime/core/providers/cpu/tensor/unique.cc

```diff
@@ -0,0 +1,354 @@
+
+#include "core/providers/cpu/tensor/unique.h"
+
+#include <map>
+#include "gsl/span"
+#include "core/framework/utils.h"
+#include "core/providers/common.h"
+
+namespace onnxruntime {
+
+/*
+ONNX_OPERATOR_SET_SCHEMA(
+    Unique,
+    11,
+    OpSchema()
+        .SetDoc(Unique_ver11_doc)
+        .Attr(
+            "sorted",
+            "(Optional) Whether to sort the unique elements in ascending order before returning as output. "
+            "Must be one of 0, or 1 (default).",
+            AttributeProto::INT,
+            static_cast<int64_t>(1))
+        .Attr(
+            "axis",
+            "(Optional) The dimension to apply unique. If not specified, the unique elements of the "
+            "flattened input are returned. Negative value means counting dimensions "
+            "from the back. Accepted range is [-r, r-1] where r = rank(input).",
+            AttributeProto::INT,
+            OPTIONAL)
+        .Input(0, "X", "A N-D input tensor that is to be processed.", "T")
+        .Output(
+            0,
+            "Y",
+            "A tensor of the same type as 'X' "
+            "containing all the unique values or subtensors sliced along a provided 'axis' in 'X', either sorted "
+            "or maintained in the same order they occur in input 'X'",
+            "T")
+        .Output(
+            1,
+            "indices",
+            "A 1-D INT64 tensor "
+            "containing indices of 'Y' elements' first occurance in 'X'. "
+            "When 'axis' is provided, it contains indices to subtensors in input 'X' on the 'axis'. "
+            "When 'axis' is not provided, it contains indices to values in the flattened input tensor. ",
+            "tensor(int64)",
+            OpSchema::Optional)
+        .Output(
+            2,
+            "inverse_indices",
+            "A 1-D INT64 tensor "
+            "containing, for elements of 'X', its corresponding indices in 'Y'. "
+            "When 'axis' is provided, it contains indices to subtensors in output 'Y' on the 'axis'. "
+            "When 'axis' is not provided, it contains indices to values in output 'Y'. ",
+            "tensor(int64)",
+            OpSchema::Optional)
+        .Output(
+            3,
+            "counts",
+            "A 1-D INT64 tensor containing "
+            "the count of each element "
+            "of 'Y' in input 'X'",
+            "tensor(int64)",
+            OpSchema::Optional)
+        .TypeConstraint(
+            "T",
+            OpSchema::all_tensor_types(),
+            "Input can be of any tensor type.")    
+*/
+ONNX_CPU_OPERATOR_KERNEL(
+    Unique,
+    11,
+    KernelDefBuilder().TypeConstraint("T", DataTypeImpl::AllTensorTypes()),
+    Unique);
+
+Status Unique::Compute(OpKernelContext* context) const {
+  const Tensor& input = *context->Input<Tensor>(0);
+
+  Status status;
+  auto data_type = input.DataType();
+
+  DispatchOnTensorTypeWithReturn(data_type, status, ComputeImpl, *context);
```

Comment:
Is supporting all the types from the get-go necessary ? I know this is a controversial subject, but it seems pertinent to raise this as there are a bunch of templated methods below that are going to be expanded for every type. Can we go with string, int32, int64, float, double, and bool (or even more conservative selection list) for now. That should cut down the types by roughly half (good savings).

Reply:
It would be good if we had a mechanism to restrict everything to a subset of types, instead of doing it on an ad hoc basis, with different ops having different types enabled. 

I can start with string, int64 and float.

Reply:
And maybe int8 in case some quantized model needs this.

Reply:
sure